spinalcordtoolbox/deepseg/inference.py

```diff
@@ -117,44 +117,39 @@ def segment_monai_single(path_img, path_out, chkp_path, crop_size="64x192x-1", d
     else:
         DEVICE = torch.device("cuda" if torch.cuda.is_available() and device == "gpu" else "cpu")
 
-    # define root path for finding datalists
+    # load model from checkpoint
     chkp_path = os.path.join(chkp_path, "best_model_loss.ckpt")
+    net = create_nnunet_from_plans(chkp_path, DEVICE)
 
     # define inference patch size and center crop size
     crop_size = tuple([int(i) for i in crop_size.split("x")])
     inference_roi_size = (64, 192, 320)
 
     # define the dataset and dataloader
     test_loader, test_post_pred = prepare_data(path_img, path_out, crop_size=crop_size)
-
-    # define model
-    net = create_nnunet_from_plans(chkp_path, device)
-
-    # iterate over the dataset and compute metrics
-    for batch in test_loader:
-        # get the test input
-        test_input = batch["image"].to(DEVICE)
-
-        # run inference
-        pred = sliding_window_inference_wrapped(
-            batch=batch,
-            test_input=test_input,
-            inference_roi_size=inference_roi_size,
-            predictor=net,
-            test_post_pred=test_post_pred
-        )
-
-        # this takes about 0.25s on average on a CPU
-        # image saver class
-        _, fname, ext = extract_fname(path_img)
-        postfix = "seg"
-        pred_saver = SaveImage(
-            output_dir=path_out, output_postfix=postfix, output_ext=ext,
-            separate_folder=False, print_log=False)
-        # save the prediction
-        fname_out = os.path.join(path_out, f"{fname}_{postfix}{ext}")
-        logger.info(f"Saving results to: {fname_out}")
-        pred_saver(pred)
+    batch = next(iter(test_loader))
```

Comment:
Again, my deepest apologies for not taking the time to write a clear extended commit message. I definitely could have done a better job at providing context.

> Or maybe we expect test_loader to have length 1 anyway? 

100%!!

While the [original script](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/e65478099d026f865b7f1d7d0082e6e9a507a744/monai/run_inference_single_image.py) seems to have been written to perform inference on multiple images, `sct_deepseg` only performs inference on a single image per call to `segment_monai`, so the `DataLoader` object will only ever have a single batch (i.e. the loop will only ever be run once).

Reply:
!!! This is very clever! Thank you for this!

> * It enforces that test_loader is an iterable with a single value, rather than silently ignoring extra values if they're present.

I was curious about what might happen if the iterable _doesn't_ have a single value, and:

```python
>>> tuple_2 = (1, 2)
>>> [batch] = tuple_2
Traceback (most recent call last):
  File "/opt/pycharm-2023.2.1/plugins/python/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
ValueError: too many values to unpack (expected 1)
```

This is a very useful built-in error message, communicating exactly what you suggested.

Reply:
In essence it's kind of like a syntactic sugar version of:

```python
expected_vals = 1
if not test_loader:
    raise ValueError(f"not enough values to unpack (expected {expected_vals}, got 0)")
elif len(test_loader) > 1:
    raise ValueError(f"too many values to unpack (expected {expected_vals})")
else:
    batch = test_loader[0]
```

Good to know about!